lib/ansible/inventory/data.py

```diff
@@ -33,7 +36,235 @@
 display = Display()
 
 
-class InventoryData(object):
+class InventoryHostManager(object):
+    """
+    Manager for inventory host relatives
+    """
+    def host_populate_ancestors(self, host_name, additions=None):
+        """
+        :type host_name: str
+        :type additions: list[str]
+        """
+        host = self.hosts[host_name]
+        if additions is None:
+            for group_name in host.groups:
+                for ancestor_name in self.group_get_ancestors(group_name):
+                    if ancestor_name not in host.groups:
+                        host.groups.append(ancestor_name)
+        else:
+            for group_name in additions:
+                if group_name not in host.groups:
+                    host.groups.append(group_name)
+
+    def host_add_group(self, host_name, group_name):
+        """
+        :type host_name: str
+        :type group_name: str
+        """
+        self.host_populate_ancestors(host_name)
```

Comment:
I have always thought that rebuilding ancestors when adding an object seemed like something that could be easily avoided. Additions happen by the inventory plugin for the most part. Between adding one host and the next host, a correct list of ancestors is almost never needed. Inventory plugins add parents, but they don't generally read ancestors. So my preferred solution would be to have a lazily evaluated `groups` property on hosts.

I also have a hard time understanding the function of this method. It I, as an inventory plugin author, call this, I would be doing it wrong. It seems that this is only ever used correctly if called while in the process of adding the host to the group via the `add_host` method or something like that. So this isn't really adding the host to the group, but it is maintaining synchronization between links from groups to hosts and from hosts to ancestors (which are called "groups").

Reply:
You may be right, I'll look into it. For the time being I just tried to keep parity of behavior and ancestor evaluation. I do feel like InventoryData is overly complicated and that there are probably simplifications I'm overlooking.

Regarding the naming, the methods I took from the host file I prefixed with `host_`, those from group.py `group_`. This was just to help me remember which were from where. Maybe I should give these clearer names (something along the lines of add_group_rel_to_host?) and move them into InventoryData? I'm not sure where they make the most sense aesthetically (I feel like these are overloading InventoryData if I move them there directly but I think this now is inelegant), but you're right that they shouldn't be advertised or at least be more clear what their purpose is for plugin authors.

Reply:
(I see now where I was wrong, these are both base classes of `InventoryData`, so that's how it has access to both methods)

In terms of big picture, when @bcoca spoke of this idea previously, my take-away was that many things that used to live in the host or group model would move to the `Inventory` or `InventoryData` model.

I like that idea, because keeping track of relationships in inventory-level objects makes it a lot easier to follow the save-point logic you're doing.

You're doing some of that here, by migrating the methods so far. The thing that strikes me is that the data are still maintained on the `Host` and `Group` objects. My prior reading of this proposal was that the relationships themselves (of what is a child of what) would move to the `InventoryData` as well. Then if someone referenced `host.groups`, it would have to obtain this from a backlink to the `InventoryData` object.

I found that to be a compelling idea because the plugins tend to operate on the inventory-scope objects, but that's still very wishy washy. For something more concrete, introducing an `InventoryDataCache` class or something to differentiate computed data versus the original data from the loading process could be very productive (computed data being the indirect ancestors in specific). Along the lines of what you said, reducing the methods available on `InventoryData` would be the most maintainable IMHO because that's what the plugin authors touch.